pkg/fake/iamapi.go

```diff
@@ -0,0 +1,146 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fake
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/iam/iamiface"
+)
+
+const ()
+
+// IAMAPIBehavior must be reset between tests otherwise tests will
+// pollute each other.
+type IAMAPIBehavior struct {
+	GetInstanceProfileBehavior            MockedFunction[iam.GetInstanceProfileInput, iam.GetInstanceProfileOutput]
+	CreateInstanceProfileBehavior         MockedFunction[iam.CreateInstanceProfileInput, iam.CreateInstanceProfileOutput]
+	DeleteInstanceProfileBehavior         MockedFunction[iam.DeleteInstanceProfileInput, iam.DeleteInstanceProfileOutput]
+	AddRoleToInstanceProfileBehavior      MockedFunction[iam.AddRoleToInstanceProfileInput, iam.AddRoleToInstanceProfileOutput]
+	RemoveRoleFromInstanceProfileBehavior MockedFunction[iam.RemoveRoleFromInstanceProfileInput, iam.RemoveRoleFromInstanceProfileOutput]
+}
+
+type IAMAPI struct {
+	sync.Mutex
+
+	iamiface.IAMAPI
+	IAMAPIBehavior
+
+	InstanceProfiles map[string]*iam.InstanceProfile
+}
+
+func NewIAMAPI() *IAMAPI {
+	return &IAMAPI{InstanceProfiles: map[string]*iam.InstanceProfile{}}
+}
+
+// Reset must be called between tests otherwise tests will pollute
+// each other.
+func (s *IAMAPI) Reset() {
+	s.GetInstanceProfileBehavior.Reset()
+	s.InstanceProfiles = map[string]*iam.InstanceProfile{}
```

Comment:
Is there a reason we shouldn't just reset all the behavior funcs?

Reply:
Miss on my side. Good catch1